meta-lmp-base/recipes-bsp/u-boot/u-boot-ostree-scr-fit/boot-common.cmd.in

```diff
@@ -1,17 +1,144 @@
+				setenv start_blk "${bootloader_s}"
+				run load_image
+				run update_image
+
+				setenv image_path "${ostree_root}/usr/lib/firmware/${bootloader2_s_image}"
+				echo "FIO: Writing ${image_path} ..."
+				setenv start_blk "${bootloader2_s}"
+				run load_image
+				run update_image
+
+				run saveenv_mmc
+				echo "FIO: doing warm reset to boot into secondary boot path..."
+				reset -w
+			fi;
+		fi
+	fi
+
+	# That means that we've validated already boot image over secondary path
+	# So we can update stable primary path with that images
+	if test "${fiovb.bootupgrade_available}" = "1" && test "${fiovb.upgrade_available}" = "0"; then
+		echo "FIO: update primary boot path with validated images ..."
+
+		setenv image_path "${ostree_root}/usr/lib/firmware/${bootloader_image}"
+		echo "FIO: Writing ${image_path} ..."
+		setenv start_blk "${bootloader}"
+		run load_image
+		run update_image
+
+		setenv image_path "${ostree_root}/usr/lib/firmware/${bootloader2_image}"
+		echo "FIO: Writing ${image_path} ..."
+		setenv start_blk "${bootloader2}"
+		run load_image
+		run update_image
+
+		run bootcmd_bootenv
+		if test -n "${board_is_closed}"; then
+			fiovb write_pvalue bootupgrade_available 0;
+			fiovb write_pvalue bootfirmware_version ${bootfirmware_version};
+		else
+			setenv bootupgrade_available 0;
+		fi
+
+		run saveenv_mmc
+		echo "FIO: reset ..."
+		reset
+	fi
+fi
+
 setenv bootcmd_load_f 'ext4load ${devtype} ${devnum}:${rootpart} ${fit_addr} "/boot"${kernel_image}'
 setenv bootcmd_tee_ovy 'if test -n ${optee_ovl_addr}; then imxtract ${fit_addr}#conf@@FIT_NODE_SEPARATOR@@${fdt_file_final} fdt@@FIT_NODE_SEPARATOR@@${fdt_file_final} ${fdt_addr}; fdt addr ${fdt_addr}; fdt resize 0x1000; fdt apply ${optee_ovl_addr}; fi'
 setenv bootcmd_run 'if test -n ${optee_ovl_addr}; then bootm ${fit_addr}#conf@@FIT_NODE_SEPARATOR@@${fdt_file_final} ${fit_addr}#conf@@FIT_NODE_SEPARATOR@@${fdt_file_final} ${fdt_addr}; else echo "loading ${fdt_file_final}"; bootm ${fit_addr}#conf@@FIT_NODE_SEPARATOR@@${fdt_file_final}; fi'
-setenv bootcmd_rollbackenv 'setenv kernel_image ${kernel_image2}; setenv bootargs ${bootargs2}'
-setenv bootcmd_set_rollback 'if test ! "${rollback}" = "1"; then setenv rollback 1; setenv upgrade_available 0; saveenv; fi'
-setenv bootostree 'run bootcmd_load_f; run bootcmd_tee_ovy; run bootcmd_run'
-setenv altbootcmd 'run bootcmd_otenv; run bootcmd_set_rollback; if test -n "${kernel_image2}"; then run bootcmd_rollbackenv; fi; run bootostree; reset'
-
-if test ! -e ${devtype} ${devnum}:${bootpart} uboot.env; then saveenv; fi
+setenv bootcmd_rollback 'if test -n "${kernel_image2}" && test "${fiovb.is_secondary_boot}" = "0" && test "${fiovb.rollback}" = "1"; then setenv kernel_image ${kernel_image2}; setenv bootargs ${bootargs2}; fi;'
 
-if test "${rollback}" = "1"; then run altbootcmd; else run bootcmd_otenv; run bootostree; if test ! "${upgrade_available}" = "1"; then setenv upgrade_available 1; saveenv; fi; reset; fi
+run saveenv_mmc
```

Comment:
This will cause the environment to be saved on every boot.

The way u-boot was avoiding that with bootcount is via upgrade_available (only saving when increasing bootcount). Shouldn't we do a similar thing here?